tests/components/gree/test_climate.py

```diff
@@ -0,0 +1,781 @@
+    await async_setup_component(hass, GREE_DOMAIN, {GREE_DOMAIN: {"climate": {}}})
+    await hass.async_block_till_done()
+
+
+async def test_discovery_called_once(hass, discovery, device):
+    """Test discovery is only ever called once."""
+    await async_setup_gree(hass)
+    assert discovery.call_count == 1
+
+    await async_setup_gree(hass)
+    assert discovery.call_count == 1
+
+
+async def test_discovery_setup(hass, discovery, device):
+    """Test setup of platform."""
+    MockDevice1 = build_device_mock(
+        name="fake-device-1", ipAddress="1.1.1.1", mac="aabbcc112233"
+    )
+    MockDevice2 = build_device_mock(
+        name="fake-device-2", ipAddress="2.2.2.2", mac="bbccdd223344"
+    )
+
+    discovery.return_value = [MockDevice1.device_info, MockDevice2.device_info]
+    device.side_effect = [MockDevice1, MockDevice2]
+
+    await async_setup_gree(hass)
+    await hass.async_block_till_done()
+    assert discovery.call_count == 1
+    assert len(hass.states.async_all(DOMAIN)) == 2
+
+
+async def test_discovery_setup_connection_error(hass, discovery, device):
+    """Test gree integration is setup."""
+    MockDevice1 = build_device_mock(name="fake-device-1")
+    MockDevice1.bind = AsyncMock(side_effect=DeviceNotBoundError)
+
+    MockDevice2 = build_device_mock(name="fake-device-2")
+    MockDevice2.bind = AsyncMock(side_effect=DeviceNotBoundError)
+
+    device.side_effect = [MockDevice1, MockDevice2]
+
+    await async_setup_gree(hass)
+    await hass.async_block_till_done()
+    assert discovery.call_count == 1
+
+    assert not hass.states.async_all(DOMAIN)
+
+
+async def test_update_connection_failure(hass, discovery, device, mock_now):
+    """Testing update hvac connection failure exception."""
+    device().update_state.side_effect = [
+        DEFAULT_MOCK,
+        DeviceTimeoutError,
+        DeviceTimeoutError,
+    ]
+
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    # First update to make the device available
+    state = hass.states.get(ENTITY_ID)
+    assert state.name == "fake-device-1"
+    assert state.state != STATE_UNAVAILABLE
+
+    next_update = mock_now + timedelta(minutes=10)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    next_update = mock_now + timedelta(minutes=15)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    # Then two more update failures to make the device unavailable
+    state = hass.states.get(ENTITY_ID)
+    assert state.name == "fake-device-1"
+    assert state.state == STATE_UNAVAILABLE
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)